### ISSUE TITLE
Upload all existing Whitehall non-attachment assets not modelled in the database to Asset Manager

### ISSUE THREAD

**Issue by chrisroos** [alphagov/asset-manager#216]
This includes everything in the `/mnt/uploads/whitehall/clean/` directory on the asset hosts that isn't:

* a Whitehall image (these are being handled in #215)
* an HMRC or Number 10 upload (these are being handled in #217)
* assets stored in government/uploads/system/uploads (these are being handled in #420)
* a Whitehall attachment (we're not migrating these at the moment)

These assets are ones that previously had a carrierwave uploader in Whitehall but that has subsequently been removed. The files themselves are still publically accessible, so need to be migrated. The list of paths and the number of assets is as follows:

```
| path                                                    | count | total size |
|---------------------------------------------------------+-------+------------|
| `system/uploads/attachment/file`                        |  6882 | 1.8 GiB    |
| `system/uploads/consultation_response_form/file`        |    19 | 4.3 MiB    |
| `system/uploads/edition_organisation_image_data/file`   | 11402 | 1.4 GiB    |
| `system/uploads/edition_world_location_image_data/file` |  5243 | 507.4 MiB  |
| `system/uploads/news_article/featuring_image`           |    10 | 4.1 MiB    |
| `system/uploads/news_article/image`                     |   112 | 31.5 MiB   |
| `system/uploads/topical_event/logo`                     |   490 | 44.7 MiB   |
```

There are a list of example assets with URLs in https://gist.github.com/chrisroos/ae70ae2bbc8a5e0e06e98b7aea08d233#example-whitehall-image-assets-on-integration.

We'll need to do this before we can remove the `PublicUploadsController` from Whitehall (see issue #213).

## Tasks

- [x] run `asset_manager:migrate_assets[system/uploads/attachment/file]` on whitehall in production
- [x] run `asset_manager:migrate_assets[system/uploads/consultation_response_form/file]` on whitehall in production
- [x] run `asset_manager:migrate_assets[system/uploads/edition_organisation_image_data/file]` on whitehall in production
- [x] run `asset_manager:migrate_assets[system/uploads/edition_world_location_image_data/file]` on whitehall in production
- [x] run `asset_manager:migrate_assets[system/uploads/news_article/featuring_image]` on whitehall in production
- [x] run `asset_manager:migrate_assets[system/uploads/news_article/image]` on whitehall in production
- [x] run `asset_manager:migrate_assets[system/uploads/topical_event/logo]` on whitehall in production
- [x] Make an nginx config change to proxy these paths to Asset Manager (see https://github.com/alphagov/govuk-puppet/pull/7124 for example)
- [x] Add rake tasks to delete these paths from NFS
- [x] Deploy the nginx change to production
- [x] Deploy `release_13250` of Whitehall to production
- [x] Check number of assets in Asset Manager matches number on file system
- [x] Run `whitehall` rake task `asset_manager:remove_attachment_file` on a `whitehall-backend` machine in `production`
- [x] Run `whitehall` rake task `asset_manager:remove_consultation_response_form_file` on a `whitehall-backend` machine in `production`
- [x] Run `whitehall` rake task `asset_manager:remove_edition_world_location_image_data_file` on a `whitehall-backend` machine in `production`
- [x] Run `whitehall` rake task `asset_manager:remove_news_article_featuring_image` on a `whitehall-backend` machine in `production`
- [x] Run `whitehall` rake task `asset_manager:remove_news_article_image` on a `whitehall-backend` machine in `production`
- [x] Run `whitehall` rake task `asset_manager:remove_topical_event_logo` on a `whitehall-backend` machine in `production`
- [x] Investigate why there are more `edition_organisation_image_data` assets in asset manager than on the file system
- [x] Run `whitehall` rake task `asset_manager:remove_edition_organisation_image_data_file` on a `whitehall-backend` machine in `production`

**Comment by floehopper** [comment]
It might be possible to work on this before we've made progress with #212, because in this case there's no model in the database and it will make no sense to use the CarrierWave uploader/storage mechanism.

**Comment by floehopper** [comment]
Since these are essentially "legacy" assets, @chrisroos & I think we should postpone working on this until as late as possible. We think we should be able to redirect a subset of URL paths under `/government/uploads` to Asset Manager (c.f. #214) to postpone needing to upload these assets to Asset Manager. So I'm going to move this issue down the list a bit.

**Comment by floehopper** [comment]
I've added a note to the description to explain that we should set `legacy_etag` and `legacy_last_modified` when creating these assets via the Asset Manager API.

**Comment by floehopper** [comment]
I've marked this as blocked due to the problem identified in [this comment](https://github.com/alphagov/asset-manager/issues/182#issuecomment-334786966).

**Comment by floehopper** [comment]
This has been unblocked as per [this comment](https://github.com/alphagov/asset-manager/pull/257#issuecomment-335834927).

**Comment by chrislo** [comment]
There's a full list in #341.

**Comment by chrisroos** [comment]
I _think_ we should just be left with the non-database assets stored in clean/system/uploads by the time we get to this issue. I think it might be easier to wait until we've deleted all the other assets stored in clean/system/uploads (see https://github.com/alphagov/asset-manager/issues/405) before migrating the remainder in this issue.

**Comment by chrislo** [comment]
The rake tasks to migrate the assets have been run in production. They completed in around 1hr30. 

<img width="1400" alt="screen shot 2018-01-31 at 10 46 37" src="https://user-images.githubusercontent.com/16707/35632267-2deb71f2-0674-11e8-9c0c-39d9b8fbd6bf.png">

There were more [GdsApi::Timeout](https://sentry.io/govuk/app-whitehall/issues/401840961/) exceptions during the migration than I've seen recently. All of the jobs affected suceeded on retry. I'm not sure what caused these timeouts.

**Comment by chrislo** [comment]
The changes to nginx have been deployed. I made a request for one of the migrated assets 

```
$ curl -v "https://assets-origin.publishing.service.gov.uk/government/uploads/system/uploads/edition_world_location_image_data/file/287/Rockets_pointing_skyward__Photo_credit_should_read_JUNG_YEON-JEAFPGetty_Images_.jpg?CRL$RANDOM" > /dev/null

> GET /government/uploads/system/uploads/edition_world_location_image_data/file/287/Rockets_pointing_skyward__Photo_credit_should_read_JUNG_YEON-JEAFPGetty_Images_.jpg?CRL2044 HTTP/1.1
> Host: assets-origin.publishing.service.gov.uk
> User-Agent: curl/7.54.0
> Accept: */*
>
< HTTP/1.1 200 OK
< Server: nginx
< Date: Thu, 01 Feb 2018 12:14:50 GMT
< Content-Type: image/jpeg
< Content-Length: 127125
< Connection: close
< Content-Disposition: inline; filename="Rockets_pointing_skyward__Photo_credit_should_read_JUNG_YEON-JEAFPGetty_Images_.jpg"
< Cache-Control: max-age=14400, public
< Accept-Ranges: bytes
< ETag: "514a3d2d-1f095"
< Last-Modified: Wed, 20 Mar 2013 22:50:21 GMT
< X-Frame-Options: SAMEORIGIN
< Strict-Transport-Security: max-age=31536000
< Access-Control-Allow-Origin: *
< Access-Control-Allow-Methods: GET, OPTIONS
< Access-Control-Allow-Headers: origin, authorization
<
```

And traced the requests in Kibana to confirm that the asset was eventually served by asset manager

<img width="1001" alt="screen shot 2018-02-01 at 07 19 47" src="https://user-images.githubusercontent.com/16707/35678176-57e8a4f0-0720-11e8-9f63-b12ede26fac9.png">

**Comment by chrisroos** [comment]
@chrislo: It'd be great to compare the number of these assets on the Whitehall part of the NFS mount to the number in the database - to give us confidence that they have all been migrated successfully before we delete the files from NFS.

**Comment by chrislo** [comment]
To check the number of assets we should run the following in a production console (`backend` machine)

```
find /data/uploads/whitehall/clean/system/uploads/attachment/file -type f | wc -l
find /data/uploads/whitehall/clean/system/uploads/consultation_response_form/file -type f | wc -l
find /data/uploads/whitehall/clean/system/uploads/edition_organisation_image_data/file -type f | wc -l
find /data/uploads/whitehall/clean/system/uploads/edition_world_location_image_data/file -type f | wc -l
find /data/uploads/whitehall/clean/system/uploads/news_article/featuring_image -type f | wc -l
find /data/uploads/whitehall/clean/system/uploads/news_article/image -type f | wc -l
find /data/uploads/whitehall/clean/system/uploads/topical_event/logo -type f | wc -l
```

And the following in an `asset_manager` rails console 

```
['attachment/file',
'consultation_response_form/file',
'edition_organisation_image_data/file',
'edition_world_location_image_data/file',
'news_article/featuring_image',
'news_article/image',
'topical_event/logo'].each do |path|
  undeleted = WhitehallAsset.where(legacy_url_path: %r(/government/uploads/system/uploads/#{path}/)).count
  deleted = WhitehallAsset.deleted.where(legacy_url_path: %r(/government/uploads/system/uploads/#{path}/)).count
  p [path, undeleted, deleted]
end
```

**Comment by chrislo** [comment]
@emmabeynon ran the above commands in production. The `find` commands returned, in order:

```
6882
19
11402
5243
10
112
490
```

while the ruby code returned

```
["attachment/file", 6882, 0]
["consultation_response_form/file", 19, 0]
["edition_organisation_image_data/file", 11556, 0]
["edition_world_location_image_data/file", 5243, 0]
["news_article/featuring_image", 10, 0]
["news_article/image", 112, 0]
["topical_event/logo", 490, 0]
```

It looks like there are more assets in `edition_organisation_image_data/file` in Asset Manager than there are on the file system. We'll proceed with deleting the rest from NFS and I'll investigate what the difference is on integration tomorrow when it syncs.

**Comment by chrislo** [comment]
I've looked at the entries in the integration database. There are 154 files which start with the wrong prefix and were created on Jan 19th. For example here's two assets with the same directory and file name created at different times where the first has the incorrect additional `/government/uploads` prefix. 

```
[Fri, 19 Jan 2018 17:27:54 UTC +00:00, "/government/uploads/government/uploads/system/uploads/edition_organisation_image_data/file/50/CAA.jpg"]
[Wed, 31 Jan 2018 15:18:38 UTC +00:00, "/government/uploads/system/uploads/edition_organisation_image_data/file/50/CAA.jpg"]
```

The 154 additional entries in the database explains the difference we're seeing above, so I'm going to ask 2ndline to run the final rake task. I'll create an issue to investigate these 154 files (although I don't think they're causing any problems).

**Comment by chrislo** [comment]
@chrisroos pointed out that we migrated some assets that begin `/government/uploads/government/uploads` in https://github.com/alphagov/asset-manager/issues/420 and this will account for the 154 additional assets that matched the regex in the ruby code above. I won't need to open an issue to investigate. 

Done!
